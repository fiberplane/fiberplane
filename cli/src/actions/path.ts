import type { Context } from "@/context";
import { text } from "@clack/prompts";

// TODO: validate path
export async function promptPath(ctx: Context) {
  try {
    const placeholder = "./spooking-honc";
    const result = await text({
      message: "Where should we create your project? (./relative-path)",
      placeholder,
      defaultValue: placeholder,
      validate: (value) => {
        if (value !== "" && value[0] !== ".") {
          return "Please enter a relative path.";
        }
        return undefined;
      },
    });

    if (typeof result === "string") {
      if (result === "") {
        ctx.path = placeholder;
      } else {
        ctx.path = result;
      }
    }

    return result;
  } catch (error) {
    return error;
  }
}
