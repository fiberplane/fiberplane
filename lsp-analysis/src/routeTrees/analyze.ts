import type { RouteTree } from "../types";

// type SimpleGraphStat = {
//   base: number;
//   total: number
//   references: Array<string>
// }

// // This is a simple/naive implementation of a graph analysis
// export function analyze(treeItems: Array<RouteTree>) {
//   // Where string is the key of the RouteTree and number is the nr of references
//   const complexityMap = new Map<string, SimpleGraphStat>();
//   for (const item of treeItems) {
//     const base = item.entries.length;
//     const references: Array<string> = [];
//     for (const entry of item.entries) {
//       if (entry.type === "ROUTE_TREE_REFERENCE") {
//         references.push(entry.targetId);
//       }
//     }
//     complexityMap.set(item.id, { base, total: base, references });
//   }

//   // Add references

// }

// interface RouteTree {
//   id: string;
//   entries: Array<RouteEntry>;
// }

// interface RouteEntry {
//   type: string;
//   targetId?: string;
// }

interface SimpleGraphStat {
  base: number;
  total: number;
  references: Array<string>;
}

export function analyze(treeItems: Array<RouteTree>): RouteTree | null {
  const complexityMap = new Map<string, SimpleGraphStat>();

  // Step 1: Initialize complexity map with base values
  for (const item of treeItems) {
    const base = item.entries.length;
    const references: Array<string> = [];
    for (const entry of item.entries) {
      if (entry.type === "ROUTE_TREE_REFERENCE" && entry.targetId) {
        references.push(entry.targetId);
      }
    }
    complexityMap.set(item.id, { base, total: base, references });
  }

  // Step 2: Resolve references and update total complexity
  function resolveReferences(itemId: string, visited: Set<string>): number {
    if (visited.has(itemId)) {
      return 0; // Circular reference detected, return 0 to avoid infinite loop
    }
    visited.add(itemId);

    const stat = complexityMap.get(itemId);
    if (!stat) {
      return 0; // Item not found, return 0
    }

    let totalComplexity = stat.base;
    for (const refId of stat.references) {
      totalComplexity += resolveReferences(refId, visited);
    }

    stat.total = totalComplexity;
    return totalComplexity;
  }

  for (const item of treeItems) {
    resolveReferences(item.id, new Set<string>());
  }

  // Step 3: Find the most complex RouteTree item
  let mostComplexItem: RouteTree | null = null;
  let maxComplexity = 0;

  for (const item of treeItems) {
    const stat = complexityMap.get(item.id);
    if (stat && stat.total > maxComplexity) {
      maxComplexity = stat.total;
      mostComplexItem = item;
    }
  }

  const keys = complexityMap.keys();
  // console.log('complexityMap', complexityMap.keys());
  const totalMap: Record<string, number> = {};
  for (const key of keys) {
    const stat = complexityMap.get(key);
    if (stat) {
      totalMap[key] = stat.total;
    }
  }
  console.log("totalMap", totalMap);
  return mostComplexItem;
}

// // Example usage
// const treeItems: Array<RouteTree> = [
//   {
//     id: "1",
//     entries: [
//       { type: "ROUTE_ENTRY" },
//       { type: "ROUTE_TREE_REFERENCE", targetId: "2" },
//     ],
//   },
//   {
//     id: "2",
//     entries: [
//       { type: "ROUTE_ENTRY" },
//       { type: "ROUTE_ENTRY" },
//       { type: "ROUTE_TREE_REFERENCE", targetId: "3" },
//     ],
//   },
//   {
//     id: "3",
//     entries: [{ type: "ROUTE_ENTRY" }],
//   },
// ];

// const mostComplexItem = analyze(treeItems);
// console.log("Most complex RouteTree item:", mostComplexItem);
