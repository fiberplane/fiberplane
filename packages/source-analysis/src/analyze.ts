import { logger } from "./logger";
import type {
  RouteTree,
  RouteTreeId,
  RouteTreeReference,
  TreeResource,
} from "./types";

interface TreeStatistic {
  baseCount: number;
  totalCount: number;
  references: Array<string>;
}

export function analyze(
  resources: Record<string, TreeResource>,
): RouteTree | null {
  const complexityMap = new Map<string, TreeStatistic>();

  // Step 1: Find all tree items and initialize complexity map with base values
  const treeItems: Array<RouteTree> = Object.keys(resources)
    .filter((key) => key.startsWith("ROUTE_TREE:"))
    .map((key) => resources[key] as RouteTree);
  for (const item of treeItems) {
    const base = item.entries.length;
    const references: Array<RouteTreeId> = [];
    for (const entry of item.entries) {
      if (entry.startsWith("ROUTE_TREE_REFERENCE:")) {
        const treeReference = resources[entry];
        if (!treeReference) {
          logger.warn("Resource not found", entry);
          continue;
        }
        references.push((treeReference as RouteTreeReference).targetId);
      }
    }
    complexityMap.set(item.id, {
      baseCount: base,
      totalCount: base,
      references,
    });
  }

  // Step 2: Resolve references and update total complexity
  function resolveReferences(itemId: string, visited: Set<string>): number {
    if (visited.has(itemId)) {
      return 0; // Circular reference detected, return 0 to avoid infinite loop
    }
    visited.add(itemId);

    const stat = complexityMap.get(itemId);
    if (!stat) {
      return 0; // Item not found, return 0
    }

    let totalComplexity = stat.baseCount;
    for (const refId of stat.references) {
      totalComplexity += resolveReferences(refId, visited);
    }

    stat.totalCount = totalComplexity;
    return totalComplexity;
  }

  for (const item of treeItems) {
    resolveReferences(item.id, new Set<string>());
  }

  // Step 3: Find the most complex RouteTree item
  let mostComplexItem: RouteTree | null = null;
  let maxComplexity = 0;

  for (const item of treeItems) {
    const stat = complexityMap.get(item.id);
    if (stat && stat.totalCount > maxComplexity) {
      maxComplexity = stat.totalCount;
      mostComplexItem = item;
    }
  }

  return mostComplexItem;
}
