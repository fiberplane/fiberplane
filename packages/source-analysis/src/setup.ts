import { RoutesMonitor } from "./RoutesMonitor";
import { type Logger, logger } from "./logger";

export function createRoutesMonitor(
  projectRoot: string,
  overrideLogger?: Logger,
): RoutesMonitor {
  if (overrideLogger) {
    logger.logger = overrideLogger;
  }

  return new RoutesMonitor(projectRoot);
}
