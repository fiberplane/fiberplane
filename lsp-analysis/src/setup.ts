import path from "node:path";
import { extractRouteTrees } from "./findHono";
import { getTsLib, startServer } from "./server";
import type { TsType } from "./types";
import { Watcher } from "./watcher";

export function setupMonitoring(projectRoot: string) {
  const ts: TsType = getTsLib(projectRoot);

  // Only watch the src directory (if it exists)
  const possibleLocation = path.resolve(path.join(projectRoot, "src"));
  const location = ts.sys.directoryExists(possibleLocation)
    ? possibleLocation
    : projectRoot;
  const watcher = new Watcher(location);

  const fileMap: Record<
    string,
    {
      version: number;
      content: string;
    }
  > = {};

  function getFileInfo(fileName: string) {
    return fileMap[fileName];
  }

  function getFileNames() {
    return Object.keys(fileMap);
  }

  const server = startServer({
    getFileInfo,
    getFileNames,
    location: projectRoot,
    ts,
  });

  watcher.on("fileAdded", (event) => {
    fileMap[event.payload.fileName] = {
      version: 0,
      content: event.payload.content,
    };
    server.getProgram()?.getSourceFile(event.payload.fileName);
  });

  watcher.on("fileUpdated", (event) => {
    fileMap[event.payload.fileName] = {
      version: fileMap[event.payload.fileName].version + 1,
      content: event.payload.changes[0].text,
    };
  });

  watcher.on("fileRemoved", (event) => {
    delete fileMap[event.payload.fileName];
  });

  return {
    watcher,
    findHonoRoutes: () => extractRouteTrees(server, ts),
    teardown: () => {
      watcher.teardown();
      server.dispose();
    },
  };
}
