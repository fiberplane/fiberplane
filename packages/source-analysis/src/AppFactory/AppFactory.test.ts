import path from "node:path";
import { assert, expect, test } from "vitest";
import { AppFactory } from ".";
import { analyze } from "../analyze";
import { setupMonitoring } from "../setup";

test("hono-factory", async () => {
  const absolutePath = path.join(__dirname, "../../test-cases/hono-factory");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(absolutePath);
  try {
    await watcher.start();
    const start = performance.now();
    const result = findHonoRoutes();
    console.log(
      `Duration for hono-factory: ${(performance.now() - start).toFixed(4)}`,
    );
    const { resourceManager } = result;
    const root = analyze(resourceManager.getResources());
    expect(root).not.toBeNull();
    assert(root !== null);
    const factory = new AppFactory(resourceManager);
    if (!root) {
      throw new Error("Root is null");
    }

    const app = factory.setRootTree(root.id);
    const request = new Request("http://localhost/", { method: "GET" });
    const response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(3);

    const history = factory.getHistory();
    expect(factory.hasVisited(root.id)).toBeTruthy();
    // Expect the app from the factory file to be visited
    expect(
      history.find((item) => item.startsWith("ROUTE_TREE:factory.ts@")),
    ).toBeTruthy();
    expect(
      history.find((item) => item.startsWith("ROUTE_ENTRY:factory.ts@")),
    ).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
  } finally {
    teardown();
  }
});

test("barrel-files", async () => {
  const absolutePath = path.join(__dirname, "../../test-cases/barrel-files");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(absolutePath);
  try {
    await watcher.start();
    const start = performance.now();
    const result = findHonoRoutes();
    console.log(
      `Duration for barrel files: ${(performance.now() - start).toFixed(4)}`,
    );
    const root = analyze(result.resourceManager.getResources());
    expect(root).not.toBeNull();
    assert(root !== null);
    const factory = new AppFactory(result.resourceManager);
    if (!root) {
      throw new Error("Root is null");
    }

    const app = factory.setRootTree(root.id);
    let request = new Request("http://localhost/", { method: "GET" });
    let response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(root.id)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
    // Reset the history
    factory.resetHistory();

    request = new Request("http://localhost/user/1", { method: "GET" });
    response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(root.id)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
  } finally {
    teardown();
  }
});

test("import-as", async () => {
  const absolutePath = path.join(__dirname, "../../test-cases/import-as");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(absolutePath);
  try {
    await watcher.start();
    const start = performance.now();
    const result = findHonoRoutes();
    console.log(
      `Duration for import-as: ${(performance.now() - start).toFixed(4)}`,
    );
    const root = analyze(result.resourceManager.getResources());
    expect(root).not.toBeNull();
    assert(root !== null);
    const factory = new AppFactory(result.resourceManager);
    if (!root) {
      throw new Error("Root is null");
    }

    const app = factory.setRootTree(root.id);
    let request = new Request("http://localhost/user/1", { method: "GET" });
    let response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(root.id)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();

    // console.log("------------------------------------")
    // console.log(factory.getFilesForHistory());
    // console.log("------------------------------------")
    // Reset the history
    factory.resetHistory();

    request = new Request("http://localhost/user/1/profile", { method: "GET" });
    response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(root.id)).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
    // console.log("------------------------------------")
    // console.log(factory.getFilesForHistory());
    // console.log("------------------------------------")
  } finally {
    teardown();
  }
});

test("goose-quotes", async () => {
  const absolutePath = path.join(
    __dirname,
    "../../../../examples/goose-quotes",
  );
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(absolutePath);
  try {
    await watcher.start();
    const start = performance.now();
    const result = findHonoRoutes();
    console.log(
      `Duration for goose-quotes files: ${(performance.now() - start).toFixed(4)}`,
    );
    const root = analyze(result.resourceManager.getResources());
    expect(root).not.toBeNull();
    assert(root !== null);
    const factory = new AppFactory(result.resourceManager);
    if (!root) {
      throw new Error("Root is null");
    }

    const app = factory.setRootTree(root.id);
    const request = new Request("http://localhost/", { method: "GET" });
    const response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(root.id)).toBeTruthy();
    const content = factory.getFilesForHistory();
    console.log(content);
    expect(content).toMatchSnapshot();
  } finally {
    teardown();
  }
});

test("goosify", async () => {
  const absolutePath = path.join(__dirname, "../../../../examples/goosify");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(absolutePath);
  try {
    await watcher.start();
    const start = performance.now();
    const result = findHonoRoutes();
    console.log(
      `Duration for goosify files: ${(performance.now() - start).toFixed(4)}`,
    );
    const root = analyze(result.resourceManager.getResources());
    expect(root).not.toBeNull();
    assert(root !== null);
    const factory = new AppFactory(result.resourceManager);
    if (!root) {
      throw new Error("Root is null");
    }

    const app = factory.setRootTree(root.id);
    let request = new Request("http://localhost/api/geese", {
      method: "GET",
    });
    let response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(root.id)).toBeTruthy();
    let content = factory.getFilesForHistory();
    console.log(content);
    expect(content).toMatchSnapshot();

    factory.resetHistory();
    request = new Request("http://localhost/api/Gans", {
      method: "GET",
    });
    response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);
    expect(factory.hasVisited(root.id)).toBeTruthy();
    content = factory.getFilesForHistory();
    console.log(content);
    expect(content).toMatchSnapshot();
  } finally {
    teardown();
  }
});
