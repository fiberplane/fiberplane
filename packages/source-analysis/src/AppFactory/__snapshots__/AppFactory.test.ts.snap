// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`barrel-files 1`] = `
"/* index.ts */
const app = new Hono();

app.get("/", (c) => c.text("Hello, Hono!")
/* EOF: index.ts */"
`;

exports[`barrel-files 2`] = `
"/* index.ts */
import { cors } from "hono/cors";
import { getUser } from "./db";
const app = new Hono();

app.get("/user/1", cors()async (c) => {
  // await getUser();
  const user = await getUser();
  return c.json(user);
}
/* EOF: index.ts */
/* db/index.ts */

export { getUser } from "./getUser"
/* EOF: db/index.ts */
/* db/getUser.ts */
import { measure } from "@fiberplane/hono-otel";
import { sleep } from "../utils";
import { User } from "../types";

const DEFAULT_USER_NAME = "John Doe"

const DEFAULT_EMAIL = "john@doe.com"

export const getUser = measure("getUser", async () => {
  await sleep(100);
  const value: User = {
    name: DEFAULT_USER_NAME,
    email: DEFAULT_EMAIL,
  };
  return value;
});
/* EOF: db/getUser.ts */
/* utils.ts */

export async function sleep(duration: number) {
  return new Promise((resolve) => setTimeout(resolve, duration));
}
/* EOF: utils.ts */
/* types.ts */

export type Email = string;

export type User = {
  name: string;
  email: Email;
};
/* EOF: types.ts */"
`;

exports[`goose-quotes 1`] = `
"/* src/index.ts */
import { shouldHonk } from "./utils";
const app = new Hono();

app.get("/", (c) => {
  const honk = shouldHonk(c.req) ? "Honk honk!" : "";
  console.log(\`Home page accessed. Honk: \${honk}\`);
  return c.text(\`Hello Goose Quotes! \${honk}\`.trim());
}
/* EOF: src/index.ts */
/* src/utils.ts */
import { HonoRequest } from "hono";

export function shouldHonk(r: HonoRequest) {
  const { shouldHonk } = r.query();
  return typeof shouldHonk !== "undefined";
}
/* EOF: src/utils.ts */"
`;

exports[`goosify 1`] = `
"/* src/index.ts */
import { drizzle } from "drizzle-orm/d1";
import * as schema from "./db";
const app = new Hono();

app.get("/api/geese", async (c) => {
  // NOTE - This is equivalent to a raw D1 query
  // const geese = await c.env.DB.prepare("SELECT * FROM geese").all();
  const db = drizzle(c.env.DB);
  const geese = await db.select().from(schema.geese);
  return c.json({ geese });
}
/* EOF: src/index.ts */
/* src/db/index.ts */

export { geese } from "./schema"
/* EOF: src/db/index.ts */
/* src/db/schema.ts */
import { sqliteTable,integer,text } from "drizzle-orm/sqlite-core";
import { sql } from "drizzle-orm";

export const geese = sqliteTable("geese", {
  id: integer("id", { mode: "number" }).primaryKey(),
  name: text("name").notNull(),
  avatar: text("avatar"),
  createdAt: text("created_at").notNull().default(sql\`(CURRENT_TIMESTAMP)\`),
  updatedAt: text("updated_at").notNull().default(sql\`(CURRENT_TIMESTAMP)\`),
});
/* EOF: src/db/schema.ts */"
`;

exports[`goosify 2`] = `
"/* src/index.ts */
const app = new Hono();

app.get("/api/Gans", async (c) => {
  const prompt = c.req.query("prompt") || "What's happenin' Gans?";

  const messages = [
    {
      role: "system",
      content:
        "You are a friendly German Gans. You speak only of Geese. You speak only in German. You are a little grumpy",
    },
    {
      role: "user",
      content: prompt,
    },
  ];
  const response = await c.env.AI.run(
    "@cf/thebloke/discolm-german-7b-v1-awq",
    // NOTE - This is an issue with the types
    // https://github.com/cloudflare/workerd/issues/2181
    { messages } as BaseAiTextGeneration["inputs"],
  );

  return c.json(response);
}
/* EOF: src/index.ts */"
`;

exports[`hono-factory 1`] = `
"/* index.ts */
const app = new Hono();
/* EOF: index.ts */
/* factory.ts */
const app = new Hono();

app.get("/", (c) => c.text("Hello, Hono!")
/* EOF: factory.ts */"
`;

exports[`import-as 1`] = `
"/* index.ts */
import { cors } from "hono/cors";
import * as db from "./db";
const app = new Hono();

app.get("/user/1", cors()async (c) => {
  const user = await db.getUser();
  return c.json(user);
}
/* EOF: index.ts */
/* db.ts */
import { measure } from "@fiberplane/hono-otel";

const sleep = (duration = 100) =>
  new Promise((resolve) => setTimeout(resolve, duration))

export const getUser = measure("getUser", async () => {
  await sleep();
  const value = {
    name: "John Doe",
    email: "john@doe.com",
  };
  return value;
});
/* EOF: db.ts */"
`;

exports[`import-as 2`] = `
"/* index.ts */
import { cors } from "hono/cors";
import { getProfile as getUserProfile } from "./db";
const app = new Hono();

app.get("/user/1/profile", cors()async (c) => {
  const profile = await getUserProfile();
  return c.json(profile);
}
/* EOF: index.ts */
/* db.ts */
import { measure } from "@fiberplane/hono-otel";

const sleep = (duration = 100) =>
  new Promise((resolve) => setTimeout(resolve, duration))

export const getUser = measure("getUser", async () => {
  await sleep();
  const value = {
    name: "John Doe",
    email: "john@doe.com",
  };
  return value;
});

export async function getProfile() {
  const user = await getUser();
  await sleep(10);
  return {
    ...user,
    image: "https://xsgames.co/randomusers/avatar.php?g=pixel",
  };
}
/* EOF: db.ts */"
`;
